File: backend/rutas/usuarios.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from dependencias import obtener_db, obtener_usuario_actual, requerir_admin
from modelos.usuario import Usuario
from modelos.sector import Sector
from esquemas.usuario import (
    UsuarioCrear, UsuarioActualizar, UsuarioSalidaConSector,
    ActualizarSectorPropio, ActualizarSectoresPropio, ActualizarPerfilPropio,
)
from seguridad import hashear_password

router = APIRouter(prefix="/usuarios", tags=["usuarios"])


def _resolver_sectores(db: Session, sector_ids: list[int]) -> list[Sector]:
    sectores = db.query(Sector).filter(Sector.id.in_(sector_ids)).all()
    if len(sectores) != len(sector_ids):
        raise HTTPException(status_code=400, detail="Uno o más sectores no existen")
    return sectores


@router.get("", response_model=list[UsuarioSalidaConSector])
def listar_usuarios(db: Session = Depends(obtener_db), _: Usuario = Depends(requerir_admin)):
    return db.query(Usuario).all()


@router.post("", response_model=UsuarioSalidaConSector, status_code=201)
def crear_usuario(datos: UsuarioCrear, db: Session = Depends(obtener_db), _: Usuario = Depends(requerir_admin)):
    if db.query(Usuario).filter(Usuario.email == datos.email).first():
        raise HTTPException(status_code=400, detail="El email ya esta registrado")
    usuario = Usuario(
        nombre=datos.nombre,
        apellido=datos.apellido,
        email=datos.email,
        password_hash=hashear_password(datos.password),
        rol=datos.rol,
    )
    if datos.sector_ids:
        usuario.sectores = _resolver_sectores(db, datos.sector_ids)
        usuario.sector_id = datos.sector_ids[0]
        usuario.seleccion_completada = True
    db.add(usuario)
    db.commit()
    db.refresh(usuario)
    return usuario


@router.get("/yo", response_model=UsuarioSalidaConSector)
def obtener_perfil_propio(
    usuario_actual: Usuario = Depends(obtener_usuario_actual),
    db: Session = Depends(obtener_db),
):
    db.refresh(usuario_actual)
    return usuario_actual


@router.put("/yo", response_model=UsuarioSalidaConSector)
def actualizar_perfil_propio(
    datos: ActualizarPerfilPropio,
    db: Session = Depends(obtener_db),
    usuario_actual: Usuario = Depends(obtener_usuario_actual),
):
    if datos.email and datos.email != usuario_actual.email:
        if db.query(Usuario).filter(Usuario.email == datos.email).first():
            raise HTTPException(status_code=400, detail="El email ya está en uso")
    for campo, valor in datos.model_dump(exclude_none=True).items():
        if campo == "password":
            setattr(usuario_actual, "password_hash", hashear_password(valor))
        else:
            setattr(usuario_actual, campo, valor)
    db.commit()
    db.refresh(usuario_actual)
    return usuario_actual


@router.put("/yo/sectores", response_model=UsuarioSalidaConSector)
def actualizar_sectores_propios(
    datos: ActualizarSectoresPropio,
    db: Session = Depends(obtener_db),
    usuario_actual: Usuario = Depends(obtener_usuario_actual),
):
    usuario_actual.sectores = _resolver_sectores(db, datos.sector_ids) if datos.sector_ids else []
    usuario_actual.ver_todos = datos.ver_todos
    # Si el sector activo ya no está en la lista y no es "ver todos", limpiarlo
    if not datos.ver_todos and usuario_actual.sector_id and usuario_actual.sector_id not in datos.sector_ids:
        usuario_actual.sector_id = datos.sector_ids[0] if datos.sector_ids else None
    db.commit()
    db.refresh(usuario_actual)
    return usuario_actual


@router.put("/yo/sector", response_model=UsuarioSalidaConSector)
def actualizar_sector_activo(
    datos: ActualizarSectorPropio,
    db: Session = Depends(obtener_db),
    usuario_actual: Usuario = Depends(obtener_usuario_actual),
):
    # Verificamos que los sectores existan
    sectores = _resolver_sectores(db, datos.sector_ids) if datos.sector_ids else []
    
    usuario_actual.sectores = sectores
    usuario_actual.ver_todos = datos.ver_todos
    usuario_actual.seleccion_completada = True
    
    # El sector_id (activo) será el primero de la lista o None
    usuario_actual.sector_id = datos.sector_ids[0] if datos.sector_ids else None
    
    db.commit()
    db.refresh(usuario_actual)
    return usuario_actual


@router.put("/{id}", response_model=UsuarioSalidaConSector)
def actualizar_usuario(
    id: int,
    datos: UsuarioActualizar,
    db: Session = Depends(obtener_db),
    _: Usuario = Depends(requerir_admin),
):
    usuario = db.query(Usuario).filter(Usuario.id == id).first()
    if not usuario:
        raise HTTPException(status_code=404, detail="Usuario no encontrado")

    campos = datos.model_dump(exclude_none=True)
    sector_ids = campos.pop("sector_ids", None)

    for campo, valor in campos.items():
        if campo == "password":
            setattr(usuario, "password_hash", hashear_password(valor))
        else:
            setattr(usuario, campo, valor)

    if sector_ids is not None:
        usuario.sectores = _resolver_sectores(db, sector_ids) if sector_ids else []
        if usuario.sector_id not in [s.id for s in usuario.sectores]:
            usuario.sector_id = sector_ids[0] if sector_ids else None
    
    if sector_ids or usuario.ver_todos:
        usuario.seleccion_completada = True

    db.commit()
    db.refresh(usuario)
    return usuario


@router.delete("/{id}", status_code=204)
def eliminar_usuario(id: int, db: Session = Depends(obtener_db), _: Usuario = Depends(requerir_admin)):
    usuario = db.query(Usuario).filter(Usuario.id == id).first()
    if not usuario:
        raise HTTPException(status_code=404, detail="Usuario no encontrado")
    usuario.activo = False
    db.commit()

File: backend-node/src/routes/usuarios.js
import { Router } from 'express';
import { prisma } from '../db.js';
import { obtenerUsuarioActual, requerirAdminOSuperior, requerirPermiso } from '../middleware/auth.js';
import { hashearPassword } from '../seguridad.js';

const router = Router();

const INCLUDE_USUARIO = {
  sector: true,
  sectores: { include: { sector: true } },
};

function serializarUsuario(u) {
  return {
    id: u.id,
    nombre: u.nombre,
    apellido: u.apellido,
    email: u.email,
    rol: u.rol,
    sector_id: u.sector_id,
    ver_todos: u.ver_todos,
    seleccion_completada: u.seleccion_completada,
    activo: u.activo,
    fecha_creacion: u.fecha_creacion,
    sector: u.sector ? { id: u.sector.id, nombre: u.sector.nombre } : null,
    sectores: (u.sectores || []).map(us => ({ id: us.sector.id, nombre: us.sector.nombre })),
  };
}

async function resolverSectores(sectorIds) {
  const sectores = await prisma.sector.findMany({ where: { id: { in: sectorIds } } });
  if (sectores.length !== sectorIds.length) {
    const err = new Error('Uno o mas sectores no existen'); err.status = 400; throw err;
  }
  return sectores;
}

// GET /api/usuarios
router.get('/', obtenerUsuarioActual, async (req, res) => {
  const usuarios = await prisma.usuario.findMany({
    where: {
      email: {
        not: {
          startsWith: 'incognito_'
        }
      }
    },
    include: INCLUDE_USUARIO
  });
  return res.json(usuarios.map(serializarUsuario));
});

// POST /api/usuarios
router.post('/', requerirPermiso('admin_usuarios'), async (req, res) => {
  const { nombre, apellido, email, password, rol = 'usuario', sector_ids = [] } = req.body;
  const existe = await prisma.usuario.findFirst({ where: { email } });
  if (existe) return res.status(400).json({ detail: 'El email ya esta registrado' });

  const data = { nombre, apellido, email, password_hash: hashearPassword(password), rol };
  if (rol === 'admin') {
    const todosSectores = await prisma.sector.findFirst({ where: { nombre: 'Todos los sectores' } });
    if (todosSectores) {
      data.sector_id = todosSectores.id;
      data.ver_todos = true;
      data.seleccion_completada = true;
      data.sectores = { create: [{ sector_id: todosSectores.id }] };
    }
  } else if (sector_ids.length > 0) {
    await resolverSectores(sector_ids);
    data.sector_id = sector_ids[0];
    data.seleccion_completada = true;
    data.sectores = { create: sector_ids.map(id => ({ sector_id: id })) };
  }

  const usuario = await prisma.usuario.create({ data, include: INCLUDE_USUARIO });
  return res.status(201).json(serializarUsuario(usuario));
});

// GET /api/usuarios/yo  (debe ir ANTES de /:id)
router.get('/yo', obtenerUsuarioActual, async (req, res) => {
  const u = await prisma.usuario.findUnique({ where: { id: req.usuario.id }, include: INCLUDE_USUARIO });
  return res.json(serializarUsuario(u));
});

// PUT /api/usuarios/yo
router.put('/yo', obtenerUsuarioActual, async (req, res) => {
  const { nombre, apellido, email, password } = req.body;
  const u = req.usuario;
  if (email && email !== u.email) {
    const existe = await prisma.usuario.findFirst({ where: { email } });
    if (existe) return res.status(400).json({ detail: 'El email ya esta en uso' });
  }
  const data = {};
  if (nombre !== undefined) data.nombre = nombre;
  if (apellido !== undefined) data.apellido = apellido;
  if (email !== undefined) data.email = email;
  if (password !== undefined) data.password_hash = hashearPassword(password);
  const actualizado = await prisma.usuario.update({ where: { id: u.id }, data, include: INCLUDE_USUARIO });
  return res.json(serializarUsuario(actualizado));
});

// PUT /api/usuarios/yo/sectores
router.put('/yo/sectores', obtenerUsuarioActual, async (req, res) => {
  const { sector_ids = [], ver_todos = false } = req.body;
  const u = req.usuario;
  if (sector_ids.length > 0) await resolverSectores(sector_ids);
  let nuevoSectorId = u.sector_id;
  if (!ver_todos && u.sector_id && !sector_ids.includes(u.sector_id)) {
    nuevoSectorId = sector_ids.length > 0 ? sector_ids[0] : null;
  }
  await prisma.usuarioSector.deleteMany({ where: { usuario_id: u.id } });
  const actualizado = await prisma.usuario.update({
    where: { id: u.id },
    data: { ver_todos, sector_id: nuevoSectorId, sectores: { create: sector_ids.map(id => ({ sector_id: id })) } },
    include: INCLUDE_USUARIO,
  });
  return res.json(serializarUsuario(actualizado));
});

// PUT /api/usuarios/yo/sector
router.put('/yo/sector', obtenerUsuarioActual, async (req, res) => {
  const { sector_ids = [], ver_todos = false } = req.body;
  const u = req.usuario;
  if (sector_ids.length > 0) await resolverSectores(sector_ids);
  await prisma.usuarioSector.deleteMany({ where: { usuario_id: u.id } });
  const actualizado = await prisma.usuario.update({
    where: { id: u.id },
    data: {
      ver_todos, seleccion_completada: true,
      sector_id: sector_ids.length > 0 ? sector_ids[0] : null,
      sectores: { create: sector_ids.map(id => ({ sector_id: id })) },
    },
    include: INCLUDE_USUARIO,
  });
  return res.json(serializarUsuario(actualizado));
});

// PUT /api/usuarios/:id
router.put('/:id', requerirPermiso('admin_usuarios'), async (req, res) => {
  const id = parseInt(req.params.id);
  const usuario = await prisma.usuario.findUnique({ where: { id } });
  if (!usuario) return res.status(404).json({ detail: 'Usuario no encontrado' });

  const { nombre, apellido, email, password, rol, sector_id, sector_ids, ver_todos, seleccion_completada, activo } = req.body;
  const data = {};
  if (nombre !== undefined) data.nombre = nombre;
  if (apellido !== undefined) data.apellido = apellido;
  if (email !== undefined) data.email = email;
  if (password !== undefined) data.password_hash = hashearPassword(password);
  if (rol !== undefined) data.rol = rol;
  if (sector_id !== undefined) data.sector_id = sector_id;
  if (ver_todos !== undefined) data.ver_todos = ver_todos;
  if (seleccion_completada !== undefined) data.seleccion_completada = seleccion_completada;
  if (activo !== undefined) data.activo = activo;

  const nuevoRol = rol !== undefined ? rol : usuario.rol;
  if (nuevoRol === 'admin') {
    const todosSectores = await prisma.sector.findFirst({ where: { nombre: 'Todos los sectores' } });
    if (todosSectores) {
      data.sector_id = todosSectores.id;
      data.ver_todos = true;
      data.seleccion_completada = true;
      await prisma.usuarioSector.deleteMany({ where: { usuario_id: id } });
      data.sectores = { create: [{ sector_id: todosSectores.id }] };
    }
  } else if (sector_ids !== undefined) {
    if (sector_ids.length > 0) await resolverSectores(sector_ids);
    await prisma.usuarioSector.deleteMany({ where: { usuario_id: id } });
    data.sectores = { create: sector_ids.map(sid => ({ sector_id: sid })) };
    const sectorActualId = data.sector_id ?? usuario.sector_id;
    if (!sector_ids.includes(sectorActualId)) {
      data.sector_id = sector_ids.length > 0 ? sector_ids[0] : null;
    }
  }

  if (nuevoRol !== 'admin' && ((sector_ids && sector_ids.length > 0) || ver_todos)) {
    data.seleccion_completada = true;
  }

  const actualizado = await prisma.usuario.update({ where: { id }, data, include: INCLUDE_USUARIO });
  return res.json(serializarUsuario(actualizado));
});

// DELETE /api/usuarios/:id
router.delete('/:id', requerirPermiso('admin_usuarios'), async (req, res) => {
  const id = parseInt(req.params.id);
  const usuario = await prisma.usuario.findUnique({ where: { id } });
  if (!usuario) return res.status(404).json({ detail: 'Usuario no encontrado' });
  await prisma.usuario.update({ where: { id }, data: { activo: false } });
  return res.status(204).send();
});

export default router;
